File: src/chapters/linear-systems/content/code/gauss-jordan.cpp
#include <vector>
#include <iostream>
#include <cmath>
using namespace std;
using Vec = vector<double>;
using Mat = vector<Vec>;

// Solve A x = b by Gauss-Jordan elimination (reduced row echelon form).
Vec gauss_jordan(Mat A, Vec b) {
    int n = b.size();
    for (int i = 0; i < n; ++i) A[i].push_back(b[i]);   // augment [A | b]
    for (int k = 0; k < n; ++k) {
        int p = k;
        for (int i = k + 1; i < n; ++i) if (fabs(A[i][k]) > fabs(A[p][k])) p = i;
        swap(A[k], A[p]);
        double d = A[k][k];
        for (int j = 0; j <= n; ++j) A[k][j] /= d;       // normalize pivot row
        for (int i = 0; i < n; ++i)
            if (i != k) {
                double f = A[i][k];
                for (int j = 0; j <= n; ++j) A[i][j] -= f * A[k][j];
            }
    }
    Vec x(n);
    for (int i = 0; i < n; ++i) x[i] = A[i][n];
    return x;
}

int main() {
    Mat A = {{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}};
    Vec b = {8, -11, -3};
    Vec x = gauss_jordan(A, b);
    cout << "x ="; for (double v : x) cout << " " << v; cout << "\n";
}

File: src/chapters/linear-systems/content/code/gauss-jordan.rs
// Solve A x = b by Gauss-Jordan elimination (reduced row echelon form).
fn gauss_jordan(a: &[Vec<f64>], b: &[f64]) -> Vec<f64> {
    let n = b.len();
    let mut m: Vec<Vec<f64>> = (0..n).map(|i| {
        let mut row = a[i].clone(); row.push(b[i]); row
    }).collect();
    for k in 0..n {
        let mut p = k;
        for i in k + 1..n { if m[i][k].abs() > m[p][k].abs() { p = i; } }
        m.swap(k, p);
        let d = m[k][k];
        for j in 0..=n { m[k][j] /= d; }
        for i in 0..n {
            if i != k {
                let f = m[i][k];
                for j in 0..=n { m[i][j] -= f * m[k][j]; }
            }
        }
    }
    (0..n).map(|i| m[i][n]).collect()
}
fn main() {
    let a = vec![vec![2.0, 1.0, -1.0], vec![-3.0, -1.0, 2.0], vec![-2.0, 1.0, 2.0]];
    let b = vec![8.0, -11.0, -3.0];
    println!("{:?}", gauss_jordan(&a, &b));
}
